website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+**Step 6**
+
+
+
+* **Continuous Monitoring**: Post-deployment, our operator continuously monitors the status of the running Flink job. This real-time feedback loop enables us to promptly address any issues that may arise, ensuring the overall health and optimal performance of our Flink applications.
+
+In summary, our deployment process leverages Apache Beam Pipeline Options, integrates seamlessly with Kubernetes and the Flink Operator, and employs custom logic to handle configuration uploads and authentication. This end-to-end workflow ensures a reliable and efficient deployment of Flink applications in Kubernetes clusters while maintaining vigilant monitoring for smooth operation. We created a sequence diagram to follow the steps below.
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/job-start-activity-diagram.png"
+alt="Job Start Activity Diagram">
+
+
+
+# Developing an Autoscaler
+
+Autoscaler is the key of the self-managed streaming service. There are not enough resources available on the internet for us to learn to build our own autoscaler which makes this problem even more tricky.
+
+Autoscaler needs to scale up the number of task managers to drain the lag and keep up with the throughput. It will also scale down the minimum number of resources required to process the incoming traffic to save cost. We need to do this frequently while keeping the processing disruption to minimum.
+
+We have tuned the autoscaler intensively, so that we can meet the SLA for latency, and make trade off for cost. We also make the autoscaler application specific so that we can meet specific needs for certain applications. Every decision has a hidden cost.  We dive into the great details of autoscaler in Part 2 of this blog.
+
+
+# Creating a Client Library for Steaming Job Development
+
+To deploy the job via Flink Kubernetes Operator, one has to be knowledgeable enough in the workings of kubernetes. The steps to create a single Flink job is:
+
+
+
+1. Define a yaml file with proper specifications. This is an example:
+
+
+<img class="center-block"
+src="/images/blog/apache-beam-flink-and-kubernetes/flink-deployment-yaml.png"
+alt="Flink Deployment Yaml">
+
+
+2. SSH inside your Flink cluster and run the command:
+
+    ```
+kubectl create -f job1.yaml
+```
+
+
+3. Check the status of the job:
+
+    ```
+kubectl get flinkdeployment job1
+```
+
+
+
+This process impacts our scalability. We frequently update our jobs so It is not feasible to do this manually for thousands of running jobs, as it's highly error prone and slow. One wrong spacing in the yaml can fail the deployment. This approach also acts as a barrier to innovation as one must know kubernetes to interact with Flink jobs.
```

Comment:
```suggestion
This process impacts our scalability. Because we frequently update our jobs, we can't manually follow these steps for every running job. To do so would be highly error prone and time consuming. One wrong space in the YAML can fail the deployment. This approach also acts as a barrier to innovation, because you need to know Kubernetes to interact with Flink jobs.
```